### ISSUE TITLE
Migration with only `txes-fn` keeps being reapplied and fails

### ISSUE THREAD

**Issue by cjohansen** [avescodes/conformity#38]
I have a migration that looks like this:

```
{:myapp/migration-2017-08-08-migrate-auth0-identities
 {:txes-fn 'myapp.db.migrate-auth0-identities/migrate}}
```

This ran fine the first time I ran it, confirmed by looking at my data, and by finding `:myapp/migration-2017-08-08-migrate-auth0-identities` in `:confirmity/conformed-norms`. However, conformity keeps running this migration, and when it produces an empty list of transactions, it fails. I would expect this migration not to be run again - am I having the wrong expectations, or is this a bug? This is with 0.5.1

**Comment by cjohansen** [comment]
There seems to be two separate issues here:

1. The already conformed norm is run over again
2. A `txes-fn` norm that produces an empty list of transactions is considered an error

We're running these migrations against multiple databases, where the migration function moves some data around - however, in some databases (the test db for one), there is no data to move, so the function returns `()`, which conformity considers an error. Any way to work around or fix this?

**Comment by rkneufeld** [comment]
Hmm, I might have to defer to @daemianmack who added this feature recently.

**Comment by daemianmack** [comment]
Thanks for reporting this! 

I _think_ there's a single problem here, rather than two.

TLDR: Migration fns should return `'(())` rather than `()`.

> 1. The already conformed norm is run over again

I suspect what you're seeing here is the error `No transactions provided for norm <YOUR-NORM>` on successive runs of Conformity for the same norm. Is that correct?

The norm is being checked for conformance, and, being an empty seq, will always fail the `pos?` [check](https://github.com/rkneufeld/conformity/blob/master/src/io/rkn/conformity.clj#L97) and thus Conformity will always attempt to apply it.

The fix here would be to have your migration fn return, rather than the empty seq (`()`), a seq containing the empty seq (`'(())` or `[[]]`).

Conformity assumes the `:txes` data structure is either non-empty or nested this way, with the rationale that you can subsequently append further transactions to a given norm and have them transacted as expected; this is undocumented. I consider this particular facility a minefield and don't recommend its use, though I'm curious if @rkneufeld has other thoughts on the topic.

> 2. A txes-fn norm that produces an empty list of transactions is considered an error

Sounds like you're running afoul of the `(empty? txes)` [check](https://github.com/rkneufeld/conformity/blob/master/src/io/rkn/conformity.clj#L175) in `reduce-norms`, having been forced down this path by the false positive resulting from item 1 above. Conformity is performing a minimal sanity check on the `:txes` data. Same root cause as above.

**Comment by cjohansen** [comment]
Yep, I had the empty case wrong. It still seems to me that the `:txes-fn` function is run every time, even for completed migrations, which seems counter-intuitive to me. This will affect start-up times in the long run. Is this intentional?

**Comment by daemianmack** [comment]
Sorry for the delay!

It's currently necessary to evaluate `:txes-fn` every time in order to consider the resulting sequence of transactions for insertion. `:txes-fn` follows the semantics of `:txes` as closely as possible. It doesn't *have* to be this way in the future -- and as you point out, it would be arguably better for certain uses cases if it didn't.

**Comment by leblowl** [comment]
@daemianmack `:txes-fn` is only evaluated so that you can verify the `tx-count` of a norm is the same as the `tx-count` of an already installed norm by same name, correct? (https://github.com/rkneufeld/conformity/blob/master/src/io/rkn/conformity.clj#L86) . Maybe I am reading the code wrong, but it also seems that if the `tx-count` of a norm doesn't match the `tx-count` of the installed version, then that norm could get re-run: https://github.com/rkneufeld/conformity/blob/master/src/io/rkn/conformity.clj#L184. Is that also correct?

**Comment by daemianmack** [comment]
@leblowl Yes -- you can later add further norms to a given norm name, and if conformity finds the count has changed relative to what's been transacted, it will transact the further norms. From my original post in this thread...

> Conformity assumes the :txes data structure is either non-empty or nested this way, with the rationale that you can subsequently append further transactions to a given norm and have them transacted as expected; this is undocumented. I consider this particular facility a minefield and don't recommend its use, though I'm curious if @rkneufeld has other thoughts on the topic.
